### PR TITLE
8301243: java/net/httpclient/http2/IdleConnectionTimeoutTest.java intermittent failure

### PR DESCRIPTION
**Pull request by c-cleary**
### Description ###
Intermittent failures of this test are observed on frequent `HttpClient` test runs. The test checks that the same connection is not used twice for two seperate requests if an Idle Connection Timeout occurs by verifying that the client-side port does not use the same port. It also verifies that when an Idle Connection Timeout does not occur, the same connection is used by verifying that the port used in both requests is the same.

The issue here is that there is no guarantee that the ports used will not be the same for when an Idle Connection Timeout occurs and so the test will/does fail intermittently.

### Summary of Changes & Justification ###
Instead of comparing the post numbers of the connections used for each request in all test cases, the connections themselves are now compared with calls to `hashCode()` like so. The connection instances themselves are accessed by using a customised `ExchangeSupplier` for the `Http2TestServer`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301243](https://bugs.openjdk.org/browse/JDK-8301243): java/net/httpclient/http2/IdleConnectionTimeoutTest.java intermittent failure

### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12457/head:pull/12457` \
`$ git checkout pull/12457`

Update a local copy of the PR: \
`$ git checkout pull/12457` \
`$ git pull https://git.openjdk.org/jdk pull/12457/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12457`

View PR using the GUI difftool: \
`$ git pr show -t 12457`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12457.diff">https://git.openjdk.org/jdk/pull/12457.diff</a>

</details>

### CODE REVIEW

File: test/jdk/java/net/httpclient/http2/IdleConnectionTimeoutTest.java [hunk 1]
```diff
@@ -154,40 +162,72 @@ private HttpResponse<String> runRequest(HttpClient hc, HttpRequest req, int slee
...
+            if (exchange instanceof TestExchangeSupplier exch) {
+                if (!acceptedFirstConnection) {
+                    firstConnectionHash = exch.getTestConnection().hashCode();
+                    acceptedFirstConnection = true;
+                    exch.sendResponseHeaders(200, 0);
+                } else {
+                    int secondConnectionHash = exch.getTestConnection().hashCode();
+                    int cmp = Integer.compare(firstConnectionHash, secondConnectionHash);
+
+                    if (cmp < 0 | cmp > 0) {
```
Comment (dfuch) [inline test/jdk/java/net/httpclient/http2/IdleConnectionTimeoutTest.java:179]: In other words, `if (cmp != 0)` ? ;-)

It may be simpler and less error prone to save the connection (as an Object) in the handler.
```
        volatile Object previousConnection = null;
```

Then the condition become:
```
                if (previousConnection == null) {
                    previousConnection = exch.getTestConnection();
                    exch.sendResponseHeaders(200, 0);
                } else {
                    var secondConnection = exch.getTestConnection();

                    if (previousConnection != secondConnection) {
                        ....
```

Comment (c-cleary) [reply test/jdk/java/net/httpclient/http2/IdleConnectionTimeoutTest.java:179]: I could be wrong but I think I might have tried that before what you're calling `previousConnection` in the sample code you've given gets set to `null` in the time between the first and second requests during the tests where a timeout fires. Which does serve to test that the connections are different to be fair but maybe for some reason `previousConnection` could be null that is unrelated to it being closed by the test server. For example `createConnection()` might close early due to a Reset Stream or Protocol Error. 

I know that's very edge case but its the reason I compared using the hash code. I'm open to changing it though of course! Might be less error prone as you say.

Also, very fair on the `if (cmp != 0)` suggestion 😅 I'll definitely shorten it to that if I stick with the comparison as it is right now.

Comment (dfuch) [reply test/jdk/java/net/httpclient/http2/IdleConnectionTimeoutTest.java:179]: I don't see how using the hash wouldn't cause the exact same issue. Also you really need volatile here, since two different requests are going to be handled by two different threads.

**Comment by c-cleary** [comment]
Changes were simplified, original means of obtaining the hash code of each connection was abandoned in favour of just comparing the object references as @dfuch suggested.

**Review by dfuch — APPROVED** [review]
LGTM! Thanks for making these changes.

**Comment by c-cleary** [comment]
As this was a minor change and testing seems to be consistent, I'll proceed with integration
